### PR TITLE
Targeted script to check for DMARC / SPF policies

### PR DESCRIPTION
**Pull request by Prince-Mendiratta**
**Details**
DMARC and SPF are a set of email authentication protocols configured domain wide which verify that incoming mail from a specific domain comes from an IP Address authorized by that domain’s administrators and help prevent email spoofing. These protocols are configured over DNS as TXT records.

Our team at [Astra Security](https://www.getastra.com) would like to contribute the scan rule to check if these protocols are configured on a domain or not.

**Screenshots**
Layer 1
---
Layer 2

Signed-off-by: Prince Mendiratta <prince.mendiratta@getastra.com>

### CODE REVIEW

File: targeted/dns-email-spoofing.js [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+ * 
+ * A function which will be invoked against a specific "targeted" message.
+ * 
+ * @param msg - the HTTP message being acted upon. This is an HttpMessage object.
+ */
+function invokeWith(msg) {
+
+    var url = msg.getRequestHeader().getURI().toString();
+    // To check if script is running
+	logger("Testing script against URL - " + url);
```
Comment (kingthorin) [inline targeted/dns-email-spoofing.js:37]: Throughout the script could you make the indents consistent, there are other really noticeable occurrences in fetchRecords as well, likely others as well.

**Review by kingthorin — COMMENTED** [review]
Other than the inconsistent indent this weeks good to me.

File: targeted/dns-email-spoofing.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	var domain = msg.getRequestHeader().getURI().getHost();
+	if(domain.startsWith("www")){
+		domain = domain.replace("www.","")
+	}
+	if (policy === "DMARC") {
+		domain = "_dmarc." + domain;
+	}
+    var scheme = "https://";
+    var path = "/resolve?name=" + domain + "&type=TXT";
+    var targetUri = scheme + providerAddress + path;
```
Comment (kingthorin) [inline targeted/dns-email-spoofing.js:119]: Scheme is only used once, why not just use the literal directly?
```suggestion
    var path = "/resolve?name=" + domain + "&type=TXT";
    var targetUri = "https://" + providerAddress + path;
```

File: targeted/dns-email-spoofing.js (hunk 1 above)
Comment (Prince-Mendiratta) [reply targeted/dns-email-spoofing.js]: Updated.

**Review by kingthorin — APPROVED** [review]
LGTM

File: targeted/dns-email-spoofing.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    return false;
+}
+
+/**
+ * Function to test regex
+ * @param {string} txtData - The string to test regex against
+ * @param {RegExp} regex   - The regex expression
+ * @returns  {boolean}	   - Returns true if match found.
+ */
+function testForMatch(txtData, regex) {
```
Comment (thc202) [inline targeted/dns-email-spoofing.js:160]: Is the function really needed?

File: targeted/dns-email-spoofing.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    if (regex.test(txtData)) {
```
Comment (thc202) [inline targeted/dns-email-spoofing.js:161]: Could return the result.

File: targeted/dns-email-spoofing.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    } else {
+        return false;
+    }
+}
+
+/**
+ * Raise an alert.
+ * @see https://www.javadoc.io/doc/org.zaproxy/zap/latest/org/parosproxy/paros/core/scanner/Alert.html
+ */
+function customAlert(pluginid, alertRisk, alertConfidence, alertName, alertDesc, alertSol, cweId, wascId, msg, url) {
```
Comment (thc202) [inline targeted/dns-email-spoofing.js:172]: e.g. `raiseAlert`

File: targeted/dns-email-spoofing.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+var HttpSender = Java.type("org.parosproxy.paros.network.HttpSender")
+var Model = Java.type("org.parosproxy.paros.model.Model")
+var HistoryReference = Java.type("org.parosproxy.paros.model.HistoryReference")
+var Control = Java.type("org.parosproxy.paros.control.Control")
+var ExtensionAlert = Java.type("org.zaproxy.zap.extension.alert.ExtensionAlert")
+var Alert = Java.type("org.parosproxy.paros.core.scanner.Alert")
+
+var session = Model.getSingleton().getSession();
+var connectionParams = Model.getSingleton().getOptionsParam().getConnectionParam();
+var extLoaderAlert = Control.getSingleton().getExtensionLoader()
```
Comment (thc202) [inline targeted/dns-email-spoofing.js:18]: `extLoader`

File: targeted/dns-email-spoofing.js (hunk 3 above)
Comment (Prince-Mendiratta) [reply targeted/dns-email-spoofing.js]: Not really, removed.

**Review by kingthorin — APPROVED** [review]

**Review by kingthorin — CHANGES_REQUESTED** [review]
One more minor thing. Just for tiddy code. Could you either remove or consistently use semi-colons?

**Comment by Prince-Mendiratta** [comment]
> One more minor thing. Just for tiddy code. Could you either remove or consistently use semi-colons?

Sure thing, bad habit :P

**Comment by kingthorin** [comment]
> Sure thing, bad habit :P

No worries, as you can tell I didn't notice it till the last second either.

**Review by kingthorin — APPROVED** [review]

**Comment by thc202** [comment]
Could you address the conflict? (Better move before one of the existing entries to avoid conflict with the other ready PR.)

**Comment by kingthorin** [comment]
@Prince-Mendiratta let us know if you need advice to get this going again.

File: CHANGELOG.md [hunk 7]
```diff
@@ -9,6 +9,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - targeted/cve-2021-22214.js > A targeted script to check for Unauthorised SSRF on GitLab - CVE 2021-22214
 - httpsender/full-session-n-csrf-nashorn.js > full session and csrf token management.
 - httpfuzzerprocessor/unexpected_responses.js > compare response codes to a (pass/fail) regex and generate alerts
+- targeted/dns-email-spoofing > Check if DMARC / SPF policies are configured on a domain.
```
Comment (kingthorin) [inline CHANGELOG.md:12]: As thc202 mentioned if you move this before one of the existing entries it won't end-up in conflict again if another PR is merged first.

**Comment by Prince-Mendiratta** [comment]
Thank you, trying this out for the first time so a bit messy. The conflict should've been resolved now I believe.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

File: CHANGELOG.md (hunk 7 above)
Comment (thc202) [reply CHANGELOG.md:12]: Merged, the other PR still has conflicts.

**Pull request merged**